src/Network/Algorand/Api/Indexer.hs

```diff
@@ -120,3 +123,61 @@ data IdxAccountResponse = IdxAccountResponse
   , iarCurrentRound :: Round
   } deriving stock (Generic, Show)
 $(deriveJSON algorandTrainOptions 'IdxAccountResponse)
+
+-- | Contains fields common to all transactions and serves as an envelope
+-- to all transactions type.
+data Transaction = Transaction
```

Comment:
Maybe let's call it `TransactionIdx` or `IdxTransaction` for example? To make it clearer that it's transaction for interaction with the indexer. Same for `Block`.

Reply:
In case of `Transaction` I still a bit confused, because node will be used to send transactions and indexer to retrieve almost all information.
What about `Block` - I thin that in future we will delete old definition.
I can name it `TransactionResp` and `BlockResp`.

Maybe it would be better to keep all this definitions (including `AccountData` and so on) at `Data` module.. not sure yet. Any way it's question of another issue / PR

Reply:
> In case of `Transaction` I still a bit confused, because node will be used to send transactions and indexer to retrieve almost all information.

I think it's ok if transaction that you send and transaction that you receive are different. I just don't like that we have two types with the same name. `TransactionResp` should be ok for now I think.

Reply:
I've found a way to delete duplicate `Transaction` - there were a lot of problems with this duplication at bridge..
so, actually we do not need it, we need only correct json instances.

Can you help? How can I derive these instances for `TransactionType`?

[data TransactionType](https://github.com/serokell/haskell-algorand-sdk/blob/8758baa9a96be18329384802b1f346e2effe80a9/src/Data/Algorand/Transaction.hs#L76-L109)

Reply:
https://developer.algorand.org/docs/reference/rest-apis/indexer/#transaction

as I understand Aeson derives `tx-type` like this:
```
{
  ...
  tx-type: {
    ...
  },
  ...
}
```

but I need this:
```
{
  ...
  tx-type: pay,
  payment-transaction: {
    ...
  },
  ...
}
```

Note, that `pay` and `payment-transaction` is example and there are different transaction types.
Is it possible to derive json instances this way?

Reply:
I think I can split this sum type to one have only type labels and others to keep special transaction data
My only concern that we use current type to build transactions

Reply:
hm...
https://developer.algorand.org/docs/features/transactions/

according to this we actually just need to put string to specify transaction type

Reply:
I can leave definition as it is and manually add json instances like this:

```
instance ToJSON TransactionType where
  toJSON txType = toJSON (value :: Text)
    where
      value = case txType of
        PaymentTransaction{} -> "pay"
        ApplicationCallTransaction{} -> "appl"
        AssetTransferTransaction{} -> "axfer"
        KeyRegistrationTransaction -> "keyreg"
        AssetConfigTransaction -> "acfg"
        AssetFreezeTransaction -> "afrz"

instance FromJSON TransactionType where
  parseJSON o = do
    value :: Text <- parseJSON o
    pure $ case value of
      "pay" -> PaymentTransaction{}
      "appl" -> ApplicationCallTransaction{}
      "axfer" -> AssetTransferTransaction{}
      "keyreg" -> KeyRegistrationTransaction
      "acfg" -> AssetConfigTransaction
      "afrz" -> AssetFreezeTransaction
      x -> error . T.unpack $ "Unmapped transaction type field name: " <> x
 ```
 
 but this way some records not initialised

Reply:
> I've found a way to delete duplicate Transaction - there were a lot of problems with this duplication at bridge..
so, actually we do not need it, we need only correct json instances.

Can you elaborate on that please? I see that previously we had
```
instance ToJSON Transaction where
  toJSON = toCanonicalJson

instance FromJSON Transaction where
  parseJSON = parseCanonicalJson
```

same for `TransactionType`. What's the problem with these instances (I can imagine they may be incorrect, but I'd like to know where the problem is and whether we can fix `toCanonicalJson` and `parseCanonicalJson` to make these things work)?

Anyway, I think we should parse `TransactionType` the same way as we do it for MsgPack. Specifically, `TransactionType` should be an object with mandatory `type` key  and other keys dependent on the `type`. Or we may have no instances for `TransactionType` and implement its encoding/decoding as part of `Transaction` encoding/decoding.
It seems impossible to construct suitable `SumEncoding` for this use case, so I would just write some boilerplate encoding/decoding code.

Reply:
For `Transaction` - in `msgpack` we use `type`, but indexer returning `tx-type` – `Could not decode response: Error in $.transactions[0]: Missing required key: type`
I think it's better to not change `msgpack` definition, because it's still relevant and such changes can break something.

For `TransactionType` I've got – `Could not decode response: Error in $.transactions[0]['tx-type']: invalid encoding for Assoc`

Reply:
In case I change `msgpack` to expect `tx-type` and keep old Json instances with canonical helper functions - it's working, but there are still missing objects, depending on `tx-type`

```json
{
    "transactions-root": "LN9CHN+zm1OHdoF2SPpftJW9YI2J/F0Zv/FJsNdAheA=",
    "transactions": [
        {
            "fee": 10000,
            "tx-type": "pay"
        }
    ],
    "seed": "sBoIA83cUpa8jNOdqmGClw85y0bYcFhfmKzrnxmhgdc=",
    "round": 1,
    "genesis-id": "testnet-v1.0",
    "rewards": {
        "rewards-level": 0,
        "rewards-pool": "7777777777777777777777777777777777777777777777777774MSJUVU",
        "rewards-residue": 250000000,
        "rewards-rate": 250000000,
        "fee-sink": "A7NMWS3NT3IUDMLVO26ULGXGIIOUQ3ND2TXSER6EBGRZNOBOUIQXHIBGDE",
        "rewards-calculation-round": 500000
    },
    "genesis-hash": "SGO1GKSzyE7IEPItTxCByw9x8FmnrCDexi9/cOUJOiI=",
    "upgrade-state": {
        "next-protocol": "https://github.com/algorand/spec/tree/22726c9dcd12d9cddce4a8bd7e8ccaa707f74101",
        "next-protocol-switch-on": 20001,
        "current-protocol": "https://github.com/algorand/spec/tree/a26ed78ed8f834e2b9ccb6eb7d3ee9f629a6e622",
        "next-protocol-approvals": 1,
        "next-protocol-vote-before": 10001
    },
    "previous-block-hash": "G8RmwKHb8iQahvMCvwJVZuI97pdN/dvSIkr7ZCI0AlM=",
    "upgrade-vote": {
        "upgrade-approve": true,
        "upgrade-delay": 0,
        "upgrade-propose": "https://github.com/algorand/spec/tree/22726c9dcd12d9cddce4a8bd7e8ccaa707f74101"
    },
    "txn-counter": 0,
    "timestamp": 1560210480
}
```

Reply:
Interesting thing, that with deriving instance for `Transaction` we have more fields:

```json
{
    ...
    "transactions": [
        {
            "sender": "GD64YIY3TWGDMCNPP553DZPPR6LDUSFQOIJVFDPPXWEG3FVOJCCDBBHU5A",
            "fee": 10000,
            "last-valid": 1000,
            "tx-type": "pay",
            "first-valid": 0
        }
    ],
    ...
```
so, I guess there are not only `type` differ in `msgpack` format

Reply:
> For Transaction - in msgpack we use `type`, but indexer returning `tx-type`

Hm, ok, what does node return if you request JSON from it? `type` or `tx-type`?

>  I think it's better to not change msgpack definition, because it's still relevant and such changes can break something.

Yes, let's not change msgpack. If JSON and msgpack are different, let's just define instances separately (without `toCanonicalJson`).

Reply:
> In case I change msgpack to expect tx-type and keep old Json instances with canonical helper functions - it's working,

But how does it work if you say that in msgpack they have `type` instead of `tx-type`?

Reply:
> Interesting thing, that with deriving instance for Transaction we have more fields:

Derived by TH or `toCanonicalJson`?

Reply:
> Hm, ok, what does node return if you request JSON from it? type or tx-type?
type (https://testnet.algoexplorerapi.io/v2/blocks/1?format=json)

> But how does it work if you say that in msgpack they have type instead of tx-type?
yes, yes. I've temporary changed `type` to `tx-type` to see what will be returned)

> Interesting thing, that with deriving instance for Transaction we have more fields
Derived by TH provide more fields than `toCanonicalJson`